### PR TITLE
[MC] Honour alignment directive fill value for non-intel

### LINKED ISSUE #30303: Assembler for arm64 ignores padding value of 0 for .balign directive
|  |  |
| --- | --- |
| Bugzilla Link | [30955](https://llvm.org/bz30955) |
| Version | 3.8 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
When a specific padding value of 0 is passed to a .balign directive in an executable area, the value is ignored and NOP instructions are inserted instead.

The value is respected for the .balignw .balignl variants. Further, if a non-zero value is specified, it's respected.

Shouldn't an explicit 0 padding value be respected for .balign?

Reproducer:
```
// Contents of foo.S
.section TestSection, "x"
  add x14, x14, 0x1
  .balign 0x10, 0
  add x14, x14, 0x1
.end
```
Command line:
```
clang -target arm64-unknown-linux -c -o foo.o foo.S
```

### PR DESCRIPTION
**Pull request by lenary**
As reported in llvm.org/PR30955, `.balign` with a fill-value of 0 did not actually align using zeroes, on non-x86 targets.

This is because the check of whether to use the code alignment routines or whether to just use the fill value was checking whether the fill value was equal to `TextAlignFillValue`, which has not been changed from its default of 0 on most targets (it has been changed for x86). However, most targets do not set the fill value because it doesn't entirely make sense -- i.e. on AArch64 there's no reasonable byte value to use for alignment, as instructions are word-sized and have to be well-aligned.

I think the check at the end `AsmParser::parseDirectiveAlign` is suspicious even on x86 - if you use `.balign <align>, 0x90` in a code section, you don't end up with a block of `0x90` repeated, you end up with a block of NOPs of various widths. This functionality is never tested.

The fix here is to modify the check to ignore the default text align fill value when choosing to do code alignment or not.

Fixes #30303

### CODE REVIEW

File: llvm/test/MC/RISCV/pr30955.s [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// RUN: llvm-mc -triple riscv32 %s -o - | FileCheck %s --check-prefix=ASM
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/pr30955.s:1]: Best to use a descriptive name instead of `PRxxxxx`.

Comment (MaskRay) [reply llvm/test/MC/RISCV/pr30955.s:1]: align-zero.s might be a good name

**Review by MaskRay — APPROVED** [review]

File: llvm/test/MC/RISCV/pr30955.s (hunk 1 above)
Comment (lenary) [reply llvm/test/MC/RISCV/pr30955.s]: I chose `align-fill-byte-zero.s` to show that it's really about byte alignment directives, and not wider alignment directives (`.{b,p2,}align{l,w}`)

Comment (MaskRay) [reply llvm/test/MC/RISCV/pr30955.s]: Sounds great!

**Comment by RSilicon** [comment]
As this involves LLVM not honoring balign, should this be backported to 19.x?

**Comment by lenary** [comment]
If you think backporting would be useful I'm happy for that to happen. I don't really want to catch the intel backend maintainers by surprise if this does have strange performance effects, but I guess they should have had some time to work out whether this has an impact or not already and they haven't commented.

**Comment by lenary** [comment]
Re-reading the [release update](https://discourse.llvm.org/t/update-on-llvm-19-x-releases/80511?u=lenary), I'm not convinced that this is an important enough bugfix for a backport - the bug has been open since 2016 without anyone else working on it.

**Comment by jmorse** [comment]
NB: we've noticed that this leads to different output binaries if you compile a C/C++ file and give clang `-via-file-asm`. It looks like x86 will usually use multi-byte-nops anyway, and this fix to the assembly parser causes the textual assembler to diverge from the built-in one.

I've uploaded #110134 to address this.

**Pull request merged**